### PR TITLE
feat(searchresultstore): adding ability to pre select variant options

### PR DESCRIPTION
**Pull request by chrisFrazier77**

### CODE REVIEW

**Review by korgon — CHANGES_REQUESTED** [review]
Can we put events into a separate file to clean up constructor (eg. configureSnapFeatures)


We decided on this name/spacing and interface:
```
searchspring.fire('controller/selectVariantOptions', {
  controllers,
  options,
});
```

Lets use explicit exact match or regex only (no string regex):
```
controllers: ['search', /^recommend_/]
```

Needs Readme.

File: packages/snap-store-mobx/src/Search/Stores/SearchResultStore.test.ts [hunk 1]
```diff
@@ -251,6 +251,38 @@ describe('SearchResultStore', () => {
 			});
 		});
 
+		it('can configured to preselect certain variants', () => {
```
Comment (korgon) [inline packages/snap-store-mobx/src/Search/Stores/SearchResultStore.test.ts:254]: Does this test actually verify that the preselected values are actually selected...?

File: packages/snap-store-mobx/src/Search/Stores/SearchResultStore.ts [hunk 2]
```diff
@@ -237,6 +237,41 @@ export class Variants {
 					selection.select(firstAvailableOption.value, true);
 				}
 			});
+		} else {
+			// select first available for each selection
```
Comment (korgon) [inline packages/snap-store-mobx/src/Search/Stores/SearchResultStore.ts:241]: This comment not quite right :|

File: packages/snap-store-mobx/src/Search/Stores/SearchResultStore.ts [hunk 3]
```diff
@@ -237,6 +237,41 @@ export class Variants {
...
+						//see if that option is in the available options
+						const availablePreferedOptions = availableOptions.find((value) => value.value.toLowerCase() == preference.toLowerCase());
+
+						//use it
+						if (availablePreferedOptions) {
+							preferencedOption = availablePreferedOptions;
+						}
+					};
+
+					if (typeof preferedOptions == 'object') {
```
Comment (korgon) [inline packages/snap-store-mobx/src/Search/Stores/SearchResultStore.ts:261]: Preferred `Array.isArray()` if that is what we are expecting.

Comment (korgon) [reply packages/snap-store-mobx/src/Search/Stores/SearchResultStore.ts:261]: Logic could be simplified if we just always convert to array, or make it require an array.

**Review by korgon — CHANGES_REQUESTED** [review]
Global events will be nice!

File: packages/snap-preact/src/setupEvents.test.tsx [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+								type: 'product',
+								variants: {
+									makeSelections: makeSelectionsSpy,
+								},
+							},
+							{
+								type: 'banner',
+								variants: {
+									makeSelections: makeSelectionsSpy,
+								},
```
[the remark this answers is not in the document]
Comment (chrisFrazier77) [reply packages/snap-preact/src/setupEvents.test.tsx:40]: shouldnt, but thats why i wanted that in the test, to make sure it handled skipping banner types correctly

File: packages/snap-preact/README.md [hunk 5]
```diff
@@ -252,9 +252,11 @@ const { search } = snap.controllers;
 ```
 
 ## properties
-
 After instantiating an instance of Snap, the following properties can be accessed. 
 
+### event-manager
```
Comment (korgon) [inline packages/snap-preact/README.md:257]: I think the property name should match up with the subheading here - `eventManager`.

File: packages/snap-store-mobx/src/Search/Stores/SearchResultStore.ts [hunk 6]
```diff
@@ -181,17 +181,17 @@ export class Variants {
...
 			this.active = variant;
 			mask.set({ mappings: this.active.mappings, attributes: this.active.attributes });
 		};
 
-		this.update(variantData, preselectedOptions);
+		this.update(variantData, config);
 	}
 
-	public update(variantData: VariantData[], preselectedOptions?: Record<string, string[]>) {
+	public update(variantData: VariantData[], config?: VariantConfig) {
```
Comment (korgon) [inline packages/snap-store-mobx/src/Search/Stores/SearchResultStore.ts:194]: If this function is called externally (in middleware for example), will the user need to pass the `config` parameter used at construction to get the same settings?

**Review by korgon — APPROVED** [review]

**Pull request merged**